python_modules/dagster/dagster/components/core/component_hierarchy.py

```diff
@@ -0,0 +1,170 @@
+    # defs
+    elif (context.path / "definitions.py").exists():
+        return DefsComponentDeclaration(path=context.path)
+    elif context.path.suffix == ".py":
+        return PythonModuleComponentDeclaration(path=context.path)
+    # folder
+    elif context.path.is_dir():
+        return build_folder_component_declaration(context)
+
+    return None
+
+
+def build_component_hierarchy(
+    context: ComponentLoadContext,
+) -> ComponentHierarchy:
+    """Builds a hierarchy of components starting from the given context.
+
+    Args:
+        context (ComponentLoadContext): The context from which to start building the hierarchy.
+
+    Returns:
+        ComponentHierarchy: The root node of the component hierarchy.
+    """
+    root_decl = build_component_declaration(context)
+    if not root_decl:
+        raise ValueError("Expected root_decl to be defined, but got None")
+    return ComponentHierarchy(root=root_decl)
+
+
+def build_root_component(
+    context: ComponentLoadContext, hierarchy: ComponentHierarchy
+) -> Optional[Component]:
+    """Builds the root component from the given context.
+
+    Args:
+        context (ComponentLoadContext): The context from which to start building the component.
+
+    Returns:
+        Optional[Component]: The root component if found, otherwise None.
+    """
+    with use_component_load_context(context):
+        return build_component_from_node(context, hierarchy.root)
+
+
+def build_component_from_node(
+    context: ComponentLoadContext, decl: ComponentDeclaration
+) -> Component:
+    """Builds a component from the given context and hierarchy node.
+
+    Args:
+        context (ComponentLoadContext): The context from which to start building the component.
+        decl (ComponentDeclaration): The node representing the component in the hierarchy.
+
+    Returns:
+        Component: The built component.
+    """
+    from dagster.components.core.defs_module import (
+        DagsterDefsComponent,
+        DefsFolderComponent,
+        load_pythonic_component,
+        load_yaml_component,
+    )
+
+    if isinstance(decl, YamlComponentDeclaration):
+        return load_yaml_component(context)
+    elif isinstance(decl, PythonicComponentDeclaration):
+        return load_pythonic_component(context)
+    elif isinstance(decl, DefsComponentDeclaration):
+        return DagsterDefsComponent(path=context.path / "definitions.py")
+    elif isinstance(decl, PythonModuleComponentDeclaration):
+        return DagsterDefsComponent(path=context.path)
+    elif isinstance(decl, FolderComponentDeclaration):
+        return DefsFolderComponent(
+            path=context.path,
+            children={
+                path: build_component_from_node(context.for_path(path), child_decl)
+                for path, child_decl in decl.children.items()
+            },
+            asset_post_processors=None,
+        )
+    else:
+        check.failed(f"Unexpected ComponentDeclaration type: {type(decl)}")
```

Comment:
is there a reason not to move the build method on to the decl classes ? seems like it would fit well

Reply:
sure

Reply:
[1]

Reply:
This is still not clear to me. Can you write the function signature of what you think should exist?